src/tests/test_upgrade_agent.py

```diff
@@ -0,0 +1,189 @@
+import datetime
+import json
+import os
+import re
+import subprocess
+from typing import List
+
+import kubernetes.client
+import kubernetes.config
+import kubernetes.dynamic
+from junit_report import JunitTestSuite
+
+from service_client import log
+from tests.base_test import BaseTest
+
+
+def skopeo(args: List[str]) -> str:
+    """
+    Runs the 'skopeo' command with the given command line arguments. Raises an exception if the command finishes with
+    a non zero exit code, or if it writes anything to the standard error stream. Returns an string containing whatever
+    the command wrote to the standard output stream.
+    """
+    process = subprocess.Popen(
+        args=["skopeo"] + args,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+    )
+    (stdout, stderr) = process.communicate()
+    if process.returncode != 0:
+        raise Exception(f"failed to run skopeo with args '{args}', finished with code {process.returncode}")
+    if len(stderr) > 0:
+        raise Exception(f"failed to run skopeo with args '{args}', wrote '{stderr}' to the standard error stream")
+    return stdout.decode("utf-8")
+
+
+def image_repo(image: str) -> str:
+    """
+    Calculates the image repository from the given image reference. For example, if the image is
+    'quay.io/my/image:v1.23' it will return 'quay.io/my/image'.
+    """
+    return re.sub(r":.*$", "", image)
+
+
+def image_digest(image: str) -> str:
+    """
+    Gets the digest of the given image.
+    """
+    result = skopeo(["inspect", "--raw", f"docker://{image}"])
+    result = json.loads(result)
+    return result["config"]["digest"]
+
+
+def repo_tags(repo: str) -> str:
+    """
+    Gets the lit of tags of the given image respository"
+    """
+    result = skopeo(["list-tags", f"docker://{repo}"])
+    result = json.loads(result)
+    return result["Tags"]
+
+
+def parse_tag(tag: str) -> List[int]:
+    """
+    Converts a tag like 'v1.2.3-45' into a list of 4 integers like [1, 2, 3, 45]. This is intended for later
+    sorting lexicographically. Returns None if the given tag isn't a version number.
+    """
+    match = re.match(r"^v(?P<x>\d+).(?P<y>\d+).(?P<z>\d+)(-(?P<r>\d+))?$", tag)
+    if match is None:
+        return None
+    x = int(match.group("x"))
+    y = int(match.group("y"))
+    z = int(match.group("z"))
+    r = match.group("r")
+    if r is not None:
+        r = int(r)
+    else:
+        r = 0
+    return [x, y, z, r]
+
+
+class TestUpgradeAgent(BaseTest):
+    @JunitTestSuite()
+    def test_upgrade_agent(self, cluster):
+        """
+        This test will prepare the cluster with the previouis agent image. Once it is ready to install it will restart
+        the service with an agent image that doesn't exist, so that all the hosts will move to `insufficient`. Then it
```

Comment:
> service with an agent image that doesn't exist...

is it still the behavior? I see you've also added upgrade from candidate images and not just a non-existing tag

Reply:
Yes, the behavior is first start the server with the previous agent image, then try to upgrade to a broken image, and then try to upgrade to the current image.